### ISSUE TITLE
loss scale going to 1e-4 and training stops

### ISSUE THREAD

**Issue by nikhil-iyer-97** [pytorch/fairseq#2488]
Hi, 

I tried running this (https://github.com/pytorch/fairseq/tree/master/examples/backtranslation) and was tinkering around with the learning rate scheduler. My loss scale goes down to 1e-4 (args.min_loss_scale) if I try to keep the lr constant after warmup for a specific number of steps ( 0.0007 lr in this case for 50000 steps; loss scale goes to zero around 52000 steps and I am training it for the standard 100000 steps). My loss and train ppl are still decreasing when this happens. Any idea how to circumvent this issue ? I have also provided my config below:  
 
 - fairseq Version : master
 - PyTorch Version : 1.4
 - OS :Linux
 - Python version: 3.7
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: 16 Tesla V100
 - Apex : 0.1(master)

**Comment by myleott** [comment]
The loss scale going down indicates you're getting a lot of very large gradients which overflow the FP16 range. If these keeps happening for many batches in a row, then training stops when min_loss_scale is reached.

One option is to set `--threshold-loss-scale=0.5` or something, which forces the loss scale to stay >= 0.5. But in the worst case this could mean training runs indefinitely, constantly throwing away overflow batches and never updating the model (i.e., num_updates stays constant).

**Comment by nikhil-iyer-97** [comment]
Thanks for the reply, I wanted to use the same peak lr used in the baseline (0.0007) to compare my approach vs the baseline. Plus, since the batch size is around 400k, reducing the lr will only hamper the performance of my approach. It would be helpful if you could suggest certain scaling laws that can be used to vary lr as a function of batch size. 

I agree the threshold loss scale argument may cause training updates to cease, though it is worth a shot. Thanks for the suggestion

**Comment by myleott** [comment]
> It would be helpful if you could suggest certain scaling laws that can be used to vary lr as a function of batch size.

Unfortunately this is challenging to do in any precise terms since it depends on the model, dataset, etc.

That said, OpenAI has a nice paper proposing some such scaling laws: https://arxiv.org/abs/2001.08361. They find in Appendix D.6 (page 25) that learning rate schedules make little difference and its the total summed learning rate over training that matters most.

If you're keeping the LR fixed at 7e-4 after the warmup, that's perhaps too aggressive and you may get better performance by keeping it fixed at 5e-4 or 6e-4 for a longer period, for example (but I haven't tried this specifically).

**Comment by nikhil-iyer-97** [comment]
Thanks a lot for the reference. Will try out the suggestions you mentioned. Thanks again for the help.
